### PR TITLE
[backend/frontend] Refactor RBAC / seggregation to put everything under the "group" (#2512)

### LINKED ISSUE #2512: Refactor RBAC / seggregation to put everything under the "group"
## Use case

Today, markings are mapped on group and role are mapped on users. This is not standard and complexify a lot the management because generally, directories (LDAP/AD) are only based on groups. The purpose of this use case is to simplify the management to put attach roles to groups and not users.

After the refactor, it should be possible to have more metadata on the role to allow default dashboards / default hidden entities.

Current situation:

* Users has roles
* Users are part of groups
* Groups are granted to markings

Modification:

* Users are part of groups (no change)
* Roles are attributed to groups (change)
* Groups are granted to markings (no change)

Add:

* Role can have a default dashboard (can be overridden by the user)
* Role can have default hidden entities (overriding the default platform settings)

### PR DESCRIPTION
**Pull request by Archidoit**
### PR content
- put roles under groups (instead of users)
- initialize the users roles with the roles of the groups they belong to
- create an overview page for groups
- create an overview page for roles
- clickable roles and groups in User/Role/Group
- roles can have default hidden entities
- platform hidden entities display : take into account the default_hidden_entities of the current user and the global hidden types
- some migrations to tsx pure function
- hidden types list in settings: indication of the entity types hidden for some roles
- remove ‘default_assignation' role attribute
- add a default group linked to the default role and with default membership = true- in role line: add the number of groups with this role
- EntityType overview > add the roles for which the entity is hidden
- modification of user-test.js with roles under groups
- backend tests modifications to take the code modifications into account
- migration file 'roles-under-groups'
- SSO mapping on roles modification

### Related issues
https://github.com/OpenCTI-Platform/opencti/issues/2512

### Notion page
https://www.notion.so/filigran/Refactor-RBAC-seggregation-to-put-everything-under-the-group-2512-0dca4304e976477bb74d1f6b68357c4d

### Figma schema
https://www.figma.com/file/Rg261SJiZKVB7COuZe8lng/Roles%2C-users-and-groups?node-id=0%3A1&t=GR6v4H6JtBNdG4tn-0

### Documentation to be added !!!!!!!
https://www.notion.so/filigran/Refactor-RBAC-seggregation-to-put-everything-under-the-group-9e08c3f27e144911b0723371edcc635e

### Remark
To be done in a future issue: Role can have a default dashboard
https://github.com/OpenCTI-Platform/opencti/issues/3002

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

File: opencti-platform/opencti-front/src/private/components/settings/entity_settings/HiddenTypesList.tsx [hunk 1]
```diff
@@ -1,14 +1,47 @@
...
 import Chip from '@mui/material/Chip';
 import { Field } from 'formik';
-import { useMutation } from 'react-relay';
+import { graphql, PreloadedQuery, useMutation, usePreloadedQuery } from 'react-relay';
+import makeStyles from '@mui/styles/makeStyles';
 import { useFormatter } from '../../../../components/i18n';
 import SelectField from '../../../../components/SelectField';
 import { entitySettingsPatch } from '../sub_types/EntitySetting';
 import useEntitySettings from '../../../../utils/hooks/useEntitySettings';
+import { RoleEditionOverview_role$data } from '../roles/__generated__/RoleEditionOverview_role.graphql';
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/entity_settings/HiddenTypesList.tsx:13]: I am not really comfortable to mix-up two behaviors in this same component.

I would suggest to split the behavior between two differents components and keep a global one for the global logic :
- < HiddenTypesListEntitySetting >
- < HiddenTypesListRoles >
- < HiddenTypesList >

What do you think ?

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+      easing: theme.transitions.easing.sharp,
+      duration: theme.transitions.duration.enteringScreen,
+    }),
+    padding: 0,
+  },
+}));
+
+const groupFragment = graphql`
+    fragment Group_group on Group {
+        id
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx:68]: Some properties are not used like standard_id

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            definition_type
+            definition
+            x_opencti_color
+            x_opencti_order
+        }
+    }
+`;
+
+const Group = ({ groupData, history }: { groupData: Group_group$key, history: History }) => {
+  const classes = useStyles();
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx:100]: You can use the hook useHistory direclty

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+      </Fab>
+      <Drawer
+        open={displayUpdate}
+        anchor="right"
+        sx={{ zIndex: 1202 }}
+        elevation={1}
+        classes={{ paper: classes.drawerPaper }}
+        onClose={handleCloseUpdate}
+      >
+        <QueryRenderer
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: Can you use a suspense here ?

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+import Checkbox from '@mui/material/Checkbox';
+import { PersonOutlined } from '@mui/icons-material';
+import { makeStyles } from '@mui/styles';
+import { PreloadedQuery } from 'react-relay/relay-hooks/EntryPointTypes';
+import { commitMutation } from '../../../../relay/environment';
+import { Theme } from '../../../../components/Theme';
+import { GroupEditionRoles_group$data } from './__generated__/GroupEditionRoles_group.graphql';
+import { GroupEditionRolesLinesSearchQuery } from './__generated__/GroupEditionRolesLinesSearchQuery.graphql';
+
+const useStyles = makeStyles<Theme>((theme) => ({
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx:17]: This styles seems not used

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    groupEditionRolesLinesSearchQuery,
+    queryRef,
+  );
+  const rolesData = (roles?.edges ?? []).map((n) => n?.node) as { id: string, name: string }[];
+
+  const groupRoles = (group.roles ?? []).map((n) => ({ id: n?.id }));
+
+  const handleToggle = (roleId?: string, groupRole?: { id: string }, event?: React.ChangeEvent<HTMLInputElement>) => {
+    if (event?.target.checked) {
+      commitMutation({
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx]: Can you use useMutation here ? and below

File: opencti-platform/opencti-front/src/private/components/settings/groups/Root.tsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      variables: { id: groupId },
+    }),
+    [groupId],
+  );
+  useSubscription(subConfig);
+  const data = usePreloadedQuery(groupQuery, queryRef);
+  const { group } = data;
+
+  return (
+    <>
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/groups/Root.tsx]: We can add a security guard

<Security needs={[SETTINGS_SETACCESSES]}>

File: opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+            id
+            name
+            description
+        }
+        default_hidden_types
+    }
+`;
+
+const Role = ({ roleData, history }: { roleData: Role_role$key, history: History }) => {
+  const classes = useStyles();
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx:114]: You can use the hook useHistory directly

File: opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+      </Fab>
+      <Drawer
+        open={displayUpdate}
+        anchor="right"
+        sx={{ zIndex: 1202 }}
+        elevation={1}
+        classes={{ paper: classes.drawerPaper }}
+        onClose={handleCloseUpdate}
+      >
+        <QueryRenderer
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx:251]: You can use a suspense here

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEdition.tsx [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    padding: '20px 20px 20px 60px',
+    backgroundColor: theme.palette.background.paper,
+  },
+  closeButton: {
+    position: 'absolute',
+    top: 12,
+    left: 5,
+    color: 'inherit',
+  },
+  importButton: {
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/roles/RoleEdition.tsx]: This class seems to be not used

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    roleEditionCapabilitiesLinesSearch,
+    queryRef,
+  );
+
+  const roleCapabilities = (role.capabilities ?? []).map((n) => ({ name: n?.name })) as { name: string }[];
+
+  const handleToggle = (capabilityId: string, event: React.ChangeEvent<HTMLInputElement>) => {
+    const roleId = role.id;
+    if (event.target.checked) {
+      commitMutation({
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx]: useMutation ? and below

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionOverview.tsx [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+const RoleEditionOverviewComponent: FunctionComponent<RoleEditionOverviewComponentProps> = ({ role, context }) => {
+  const { t } = useFormatter();
+  const initialValues = R.pick(
+    ['name', 'description'],
+    role,
+  );
+  const handleChangeFocus = (name: string) => {
+    commitMutation({
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionOverview.tsx]: useMutation ? and below

File: opencti-platform/opencti-front/src/private/components/settings/roles/Root.tsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+interface RootRoleComponentProps {
+  queryRef: PreloadedQuery<RootRoleQuery>,
+}
+
+const RootRoleComponent: FunctionComponent<RootRoleComponentProps> = ({ queryRef }) => {
+  const data = usePreloadedQuery(roleQuery, queryRef);
+  const { role } = data;
+
+  return (
+    <>
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-front/src/private/components/settings/roles/Root.tsx]: You can add security guard here ?

File: opencti-platform/opencti-graphql/src/initialization.js [hunk 14]
```diff
@@ -253,10 +253,9 @@ export const createBasicRolesAndCapabilities = async (context) => {
   await createCapabilities(context, CAPABILITIES);
   // Create roles
   await addRole(context, SYSTEM_USER, {
-    name: ROLE_DEFAULT,
-    description: 'Default role associated to all users',
+    name: ROLE_KNOWLEDGE,
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/initialization.js]: why this change? to be clearer on the naming of the roles?

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js [hunk 15]
```diff
@@ -419,7 +419,7 @@ describe('Elasticsearch pagination', () => {
     let data = await elPaginate(testContext, ADMIN_USER, READ_ENTITIES_INDICES, { search: 'malicious' });
     expect(data.edges.length).toEqual(21);
     data = await elPaginate(testContext, ADMIN_USER, READ_ENTITIES_INDICES, { search: 'with malicious' });
-    expect(data.edges.length).toEqual(45);
+    expect(data.edges.length).toEqual(46);
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: Why we have an increment here ?

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js [hunk 16]
```diff
@@ -531,7 +531,7 @@ describe('Elasticsearch pagination', () => {
     let data = await elPaginate(testContext, ADMIN_USER, READ_RELATIONSHIPS_INDICES);
     expect(data).not.toBeNull();
     const groupByIndices = R.groupBy((e) => e.node._index, data.edges);
-    expect(groupByIndices[`${ES_INDEX_PREFIX}_internal_relationships-000001`].length).toEqual(14);
+    expect(groupByIndices[`${ES_INDEX_PREFIX}_internal_relationships-000001`].length).toEqual(13);
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: Why we have a decrement here ?

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx (hunk 2 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx:68]: fixed

**Comment by RomuDeuxfois** [comment]
Suggestions :
- On access, right menu, maybe we can have menus in hierarchical order users > groups > roles or roles > groups > users

Bug founds :
- Missing I18n
image

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx (hunk 3 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx:100]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx (hunk 5 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx:17]: fixed

File: opencti-platform/opencti-graphql/src/initialization.js (hunk 14 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/initialization.js]: yes, it's because it's no more a default role, but a role with knowledge capabilities

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 16 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: 13 instead of 14 because no more relation between user and role --> the relationship 'has-role' is not counted anymore

Comment (Archidoit) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: (I have checked that by printing the 14 previous inputs and the 13 new ones)

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 15 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: 46 instead of 45 because of the new role Knowledge with description 'Role with knowledge capabilities' that is know counted because it contains 'with'

File: opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx (hunk 8 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx:114]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEdition.tsx (hunk 10 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/RoleEdition.tsx]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx (hunk 6 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/groups/GroupEditionRoles.tsx]: fixed

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 16 above)
Comment (RomuDeuxfois) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: that's much clearer, thank you!

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 15 above)
Comment (RomuDeuxfois) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: that's much clearer, thank you!

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx (hunk 11 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionOverview.tsx (hunk 12 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionOverview.tsx]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx (hunk 9 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/Role.tsx:251]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/groups/Root.tsx (hunk 7 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Root.tsx]: mmm yes... but which one? did you think of one in particular? ;)

File: opencti-platform/opencti-front/src/private/components/settings/entity_settings/HiddenTypesList.tsx (hunk 1 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/entity_settings/HiddenTypesList.tsx:13]: The 2 components are really similar... I think it would create 2 almost-identical components... But we can discuss about that :)

File: opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx (hunk 4 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Group.tsx]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/groups/Root.tsx (hunk 7 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/groups/Root.tsx]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/roles/Root.tsx (hunk 13 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/Root.tsx]: fixed

File: opencti-platform/opencti-graphql/src/domain/user.js [hunk 17]
```diff
@@ -153,9 +153,14 @@ const getUserAndGlobalMarkings = async (context, userId, userGroups, capabilitie
   return { user: computedMarkings, all: markings };
 };
 
-export const getCapabilities = async (context, userId, isUserPlatform) => {
-  const roles = await listThroughGetTo(context, SYSTEM_USER, userId, RELATION_HAS_ROLE, ENTITY_TYPE_ROLE);
-  const roleIds = roles.map((r) => r.id);
+export const getRoles = async (context, userId, userGroups) => {
```
Comment (SouadHadjiat) [inline opencti-platform/opencti-graphql/src/domain/user.js]: `userId` parameter is unused in this method

Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/domain/user.js]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/roles/CapabilitiesList.tsx [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  const { t } = useFormatter();
+  const roleCapabilities = (role.capabilities ?? []).map((n) => ({ name: n?.name })) as { name: string }[];
+  const { capabilities } = usePreloadedQuery<RoleEditionCapabilitiesLinesSearchQuery>(
+    roleEditionCapabilitiesLinesSearch,
+    queryRef,
+  );
+
+  return (
+    <List>
+      {/* eslint-disable-next-line consistent-return,array-callback-return */}
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/private/components/settings/roles/CapabilitiesList.tsx]: if think your function should return <></> if not checked

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+            sx={{
+              paddingLeft: 0,
+              backgroundColor: 'transparent',
+            }}
+          >
+            {t('Capabilities')}
+          </ListSubheader>
+        }
+      >
+        {/* eslint-disable-next-line array-callback-return,consistent-return */}
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx]: Same remark about <></>

File: opencti-platform/opencti-front/src/private/components/settings/users/User.js [hunk 20]
```diff
@@ -340,12 +343,14 @@ class UserComponent extends Component {
                     {t('Roles')}
                   </Typography>
                   <List>
-                    {user.roles.map((role) => (
+                    {R.uniq(user.roles ?? []).map((role) => (
```
Comment (richard-julien) [inline opencti-platform/opencti-front/src/private/components/settings/users/User.js]: Add a sort by name?

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql [hunk 21]
```diff
@@ -1166,8 +1167,9 @@ type User implements  BasicObject & InternalObject {
   language: String
   external: Boolean
   dashboard: String
-  roles: [Role]!
+  roles: [Role]
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/config/schema/opencti.graphql]: Strange modification, null should not be possible here

File: opencti-platform/opencti-graphql/src/config/providers.js [hunk 22]
```diff
@@ -165,6 +165,11 @@ for (let i = 0; i < providerKeys.length; i += 1) {
         const lastname = user[mappedConfig.lastname_attribute] || '';
         // region roles mapping
         const isRoleBaseAccess = isNotEmptyField(mappedConfig.roles_management);
+        if (isRoleBaseAccess) {
+          logApp.error('Warning: SSO mapping on roles is deprecated, you should clean roles_management in your config and bind on groups.');
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/config/providers.js]: You can also add that it will be removed in 5.8.0+

File: opencti-platform/opencti-front/src/private/components/settings/roles/CapabilitiesList.tsx (hunk 18 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/CapabilitiesList.tsx]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx (hunk 19 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/roles/RoleEditionCapabilities.tsx]: fixed

File: opencti-platform/opencti-graphql/config/schema/opencti.graphql (hunk 21 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/config/schema/opencti.graphql]: fixed

File: opencti-platform/opencti-graphql/src/config/providers.js (hunk 22 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/config/providers.js]: fixed

File: opencti-platform/opencti-graphql/src/domain/user.js [hunk 23]
```diff
@@ -169,6 +174,13 @@ export const batchRoleCapabilities = async (context, user, roleId) => {
   return batchListThroughGetTo(context, user, roleId, RELATION_HAS_CAPABILITY, ENTITY_TYPE_CAPABILITY, { paginate: false });
 };
 
+export const getDefaultHiddenTypes = async (context, userId, userRoles) => {
+  let userDefaultHiddenTypes = [];
+  userRoles.map((role) => role.default_hidden_types?.map((type) => userDefaultHiddenTypes.push(type)));
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/domain/user.js]: Use a map to add something in an array is a bit weird.  Your userDefaultHiddenTypes is just the result of the first map, no need to repush

File: opencti-platform/opencti-graphql/src/domain/user.js [hunk 24]
```diff
@@ -641,13 +629,17 @@ const buildCompleteUser = async (context, client) => {
   const allowed_organizations = await batchOrganizations(context, SYSTEM_USER, client.id, batchOpts);
   const userOrganizations = allowed_organizations.map((m) => m.internal_id);
   const isUserPlatform = settings.platform_organization ? userOrganizations.includes(settings.platform_organization) : true;
-  const capabilities = await getCapabilities(context, client.id, isUserPlatform);
   const [individuals, organizations, groups] = await Promise.all([individualsPromise, organizationsPromise, userGroupsPromise]);
+  const roles = await getRoles(context, groups);
+  const capabilities = await getCapabilities(context, client.id, roles, isUserPlatform);
+  const default_hidden_types = await getDefaultHiddenTypes(context, client.id, roles);
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/domain/user.js]: default hidden types and capabilities could be fetched at the same time

File: opencti-platform/opencti-graphql/src/initialization.js (hunk 14 above)
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/initialization.js]: We need to have a global talk about what to do with the default role

File: opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js [hunk 25]
```diff
@@ -0,0 +1,45 @@
+import { executionContext, SYSTEM_USER } from '../utils/access';
+import { logApp } from '../config/conf';
+import { findAll, findRoleById, userAddRelation, userIdDeleteRelation } from '../domain/user';
+import { addGroup, groupAddRelation } from '../domain/group';
+
+export const up = async (next) => {
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js]: Migration should be done in 3 steps;

1. Create group from role and create the relation
2. Iter on user and assign them on the right group
3. Delete all relations from user to roles

+ Migration name must be changed to go after the last one got from the rebase

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 15 above)
Comment (richard-julien) [inline opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: Can you explain why?

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js [hunk 26]
```diff
@@ -543,14 +543,14 @@ describe('Elasticsearch pagination', () => {
     expect(metaByEntityType['external-reference'].length).toEqual(7);
     expect(metaByEntityType['object-marking'].length).toEqual(21);
     expect(metaByEntityType['kill-chain-phase'].length).toEqual(3);
-    expect(data.edges.length).toEqual(161);
+    expect(data.edges.length).toEqual(160);
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: Can you explain why?

File: opencti-platform/opencti-graphql/src/initialization.js (hunk 14 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/initialization.js]: 1. Keep default role
2. Add a default group linked to default role + Default membership = true

File: opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js (hunk 25 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js]: + Auto assign on role must create a group with auto assign

File: opencti-platform/opencti-graphql/src/domain/user.js [hunk 27]
```diff
@@ -153,9 +153,14 @@ const getUserAndGlobalMarkings = async (context, userId, userGroups, capabilitie
...
+export const getRoles = async (context, userGroups) => {
+  const groupIds = userGroups.map((r) => r.id);
+  const roles = await listThroughGetTo(context, SYSTEM_USER, groupIds, RELATION_HAS_ROLE, ENTITY_TYPE_ROLE);
+  return roles;
+};
+
+export const getCapabilities = async (context, userId, userRoles, isUserPlatform) => {
+  const roleIds = userRoles.map((r) => r.id);
   const capabilities = await listThroughGetTo(context, SYSTEM_USER, roleIds, RELATION_HAS_CAPABILITY, ENTITY_TYPE_CAPABILITY);
   if (userId === OPENCTI_ADMIN_UUID && !R.find(R.propEq('name', BYPASS))(capabilities)) {
```
Comment (richard-julien) [inline opencti-platform/opencti-graphql/src/domain/user.js:211]: Maybe a uniq is needed on get capabilities as now duplicate can happen if user is part of groups that of roles with same capabilities

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 23 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/domain/user.js]: not really because the result of the first map is an array of array... and I want all the types to be in a simple array

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 24 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/domain/user.js:635]: fixed

File: opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js (hunk 25 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js]: fixed
issue remaining if a role and a group have the same name before the migration

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 24 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/domain/user.js]: they can't be resolved at the same time because capabilities is either a Promise either a ValueOfUnion... and default_hidden_types is a Promise

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 15 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: 46 instead of 45 because of the new role Knowledge with description 'Role WITH knowledge capabilities'

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 26 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: // 150 instead of 151 before the merge (now 160 instead of 161) because of a relationship 'has-role' that is not counted anymore (has also been removed in the count of internal-relationships)

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 27 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/domain/user.js:211]: fixed

File: opencti-platform/opencti-front/src/private/components/settings/users/User.js (hunk 20 above)
Comment (Archidoit) [reply opencti-platform/opencti-front/src/private/components/settings/users/User.js]: fixed

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 23 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/domain/user.js]: You can use .flat() function for that

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 24 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/domain/user.js]: You can get a promise for both and do a Promise.all

File: opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js (hunk 25 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js]: Hum, interesting point, it could lead to strange behavior, must be added in documentation

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 15 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: It should not be a new role because its just a rename of the default no?

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 26 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: I think its due to a missing default group with default role.

File: opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js (hunk 15 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: no, I also changed the description, and there is the word 'with' inside the description, so now the role is taking into account in the counting with the filtering 'with malicious' (presence of the world 'with' in the description).

Comment (Archidoit) [reply opencti-platform/opencti-graphql/tests/02-integration/01-database/elasticSearch-test.js]: but the changes will be removed since we keep the default role now

File: opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js (hunk 25 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/migrations/1678266052390-roles-under-groups.js]: done

File: opencti-platform/opencti-graphql/src/initialization.js (hunk 14 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/initialization.js]: done

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 23 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/domain/user.js]: fixed

File: opencti-platform/opencti-graphql/src/domain/user.js (hunk 24 above)
Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/domain/user.js]: fixed

**Review by RomuDeuxfois — APPROVED** [review]
